network-area-diagram-viewer/src/main/resources/svg.js

```diff
@@ -0,0 +1,448 @@
+
+window.addEventListener('load', function() {
+    var svgDiagram = document.getElementsByTagName("svg")[0];
+    if (svgDiagram != null) {
+        makeDraggableSvg(svgDiagram);
+    }
+});
+
+// Draggable SVG Elements from
+// https://www.petercollingridge.co.uk/tutorials/svg/interactive/dragging/
+function makeDraggableSvg(svg) {
+    var selectedElement = false, offset, transform;
+    var diagram = new Diagram(svg);
+
+    svg.addEventListener('mousedown', startDrag);
+    svg.addEventListener('mousemove', drag);
+    svg.addEventListener('mouseup', endDrag);
+    svg.addEventListener('mouseleave', endDrag);
+
+    function getMousePosition(event) {
+        // From screen coordinates to SVG coordinate system
+        var CTM = svg.getScreenCTM();
+        return {
+            x: (event.clientX - CTM.e) / CTM.a,
+            y: (event.clientY - CTM.f) / CTM.d
+        };
+    }
+
+    function startDrag(event) {
+        var draggableElem = diagram.getDraggableFrom(event);
+        if (!draggableElem) {
+            return;
+        }
+        selectedElement = draggableElem;
+        selectedElement.style.cursor = 'grabbing';
+
+        offset = getMousePosition(event);
+        var transforms = getTransformsEnsuringFirstIsTranslation(svg, selectedElement);
+        // Get initial translation amount
+        transform = transforms.getItem(0);
+        offset.x -= transform.matrix.e;
+        offset.y -= transform.matrix.f;
+    }
+
+    const CONTINUOUS_UPDATE = true;
+    function drag(event) {
+        if (selectedElement) {
+            event.preventDefault();
+            var position = getMousePosition(event);
+            var translation = {x: position.x - offset.x, y: position.y - offset.y};
+            transform.setTranslate(translation.x, translation.y);
+            if (CONTINUOUS_UPDATE) {
+                diagram.update(selectedElement, position, translation);
+            }
+        }
+    }
+
+    function endDrag(event) {
+        if (selectedElement) {
+            console.log("endDrag for element " + selectedElement.id);
+            selectedElement.style.cursor = 'grab';
+            // Desired position for the selected element is current mouse position in SVG coordinate system
+            var position = getMousePosition(event);
+            var translation = {x: position.x - offset.x, y: position.y - offset.y};
+            diagram.update(selectedElement, position, translation);
+        }
+        selectedElement = null;
+    }
+}
+
+// PowSyBl Diagram
+
+function Diagram(svg) {
+    this.getDraggableFrom = getDraggableFrom;
+    this.update = update;
+
+    function getDraggableFrom(event) {
+        if (isDraggable(event.target)) {
+            return event.target;
+        } else if (isDraggable(event.target.parentElement)) {
+            return event.target.parentElement;
+        }
+    }
+
+    function isDraggable(element) {
+        return hasId(element);
```

Comment:
We don't want that any element with an id is draggable, but only the voltage level nodes (so far). Couldn't you filter this by class attribute? Or maybe only add listener on those elements in the first place, instead of putting a listener on everything and then filtering out.

Reply:
Event listeners are added only to the main svg component. I've added a filter using allowed class names (bus nodes and boundary nodes)

Reply:
why don't you put the `mousedown` listener only on the `nad-vl-nodes` element? everything outside this group doesn't need to be clickable. And then there's no filter needed.

Reply:
I prefer to keep the code in javascript `Diagram` class relatively generic, so we could extend later what we consider draggable.